destination_middleware.go

```diff
@@ -0,0 +1,236 @@
+// same context, otherwise it will return a new context with the stored value.
+// This is used to signal to destinationPluginAdapter if the Destination is
+// wrapped into DestinationWithBatch middleware.
+func (DestinationWithBatch) setBatchEnabled(ctx context.Context, enabled bool) context.Context {
+	flag, ok := ctx.Value(ctxKeyBatchEnabled{}).(*bool)
+	if ok {
+		*flag = enabled
+	} else {
+		ctx = context.WithValue(ctx, ctxKeyBatchEnabled{}, &enabled)
+	}
+	return ctx
+}
+func (DestinationWithBatch) getBatchEnabled(ctx context.Context) bool {
+	flag, ok := ctx.Value(ctxKeyBatchEnabled{}).(*bool)
+	if !ok {
+		return false
+	}
+	return *flag
+}
+
+type destinationWithBatch struct {
+	Destination
+	defaults DestinationWithBatch
+}
+
+func (d *destinationWithBatch) Parameters() map[string]Parameter {
+	return mergeParameters(d.Destination.Parameters(), map[string]Parameter{
+		configDestinationBatchSize: {
+			Default:     strconv.Itoa(d.defaults.DefaultBatchSize),
+			Required:    false,
+			Description: "Maximum size of batch before it gets written to the destination.",
+		},
+		configDestinationBatchDelay: {
+			Default:     d.defaults.DefaultBatchDelay.String(),
+			Required:    false,
+			Description: "Maximum delay before an incomplete batch is written to the destination.",
+		},
+	})
+}
+
+func (d *destinationWithBatch) Configure(ctx context.Context, config map[string]string) error {
+	// Batching is actually implemented in the plugin adapter because it is the
+	// only place we have access to acknowledgments.
+	// We need to signal back to the adapter that batching is enabled. We do
+	// this by changing a pointer that is stored in the context. It's a bit
+	// hacky, but the only way to propagate a value back to the adapter without
+	// changing the interface.
+	d.defaults.setBatchEnabled(ctx, true)
+
+	// set defaults in the config, they will be visible to the caller as well
+	if config[configDestinationBatchSize] == "" {
+		config[configDestinationBatchSize] = strconv.Itoa(d.defaults.DefaultBatchSize)
+	}
+	if config[configDestinationBatchDelay] == "" {
+		config[configDestinationBatchDelay] = d.defaults.DefaultBatchDelay.String()
+	}
+
+	return d.Destination.Configure(ctx, config)
+}
+
+// -- DestinationWithRateLimit -------------------------------------------------
+
+const (
+	configDestinationRatePerSecond = "sdk.rate.perSecond"
+	configDestinationRateBurst     = "sdk.rate.burst"
+)
+
+// DestinationWithRateLimit adds support for rate limiting to the destination.
+// It adds two parameters to the destination config:
+//   - `sdk.rate.perSecond` - Maximum times the Write function can be called per
+//     second (0 means no rate limit).
+//   - `sdk.rate.burst` - Allow bursts of at most X writes (0 means that bursts
+//     are not allowed).
+// To change the defaults of these parameters use the fields of this struct.
+type DestinationWithRateLimit struct {
```

Comment:
If there are any concerns about `DestinationWithRateLimit` or `DestinationWithBatch` I can easily move those to separate PRs so this one is not blocked. I added them just because I needed to figure out if the middleware approach works at all.

Reply:
I do have some questions, maybe suggestions too, but I'm fine with leaving the PR as it is. If we find anything worth changing, we can do it separately to get this PR out.:)